pkg/predictor_source/inferenceservice_registry.go

```diff
@@ -84,31 +86,50 @@ func processInferenceServiceStorage(inferenceService *v1beta1.InferenceService,
 		if storageSpec != nil && storageSpec.Path != nil {
 			return "", nil, "", nil, fmt.Errorf("the InferenceService %v cannot have both the storageUri and the storage.path", nname)
 		}
+
 		u, err := url.Parse(*storageUri)
 		if err != nil || u.Scheme != "s3" {
 			return "", nil, "", nil, err
 		}
-		bucket = &u.Host
+		// TODO: Support StorageURI for other types of storage too
 		modelPath = u.Path
+		uriParameters["type"] = "s3"
+		uriParameters["bucket"] = u.Host
 	}
+
+	var storageSpecParameters map[string]string
 	if storageSpec != nil {
 		if storageSpec.StorageKey != nil {
 			secretKey = *storageSpec.StorageKey
 		}
 		if storageSpec.Parameters != nil {
-			for k, v := range *storageSpec.Parameters {
-				if k == "bucket" {
-					bucket = &v
-				}
-			}
+			storageSpecParameters = *storageSpec.Parameters
 		}
 		if storageSpec.SchemaPath != nil {
 			schemaPath = storageSpec.SchemaPath
 		}
 	}
+
+	// resolve the parameters, URI parameters taking precedence
+	if storageSpecParameters != nil {
+		parameters = storageSpecParameters
+		for k, v := range uriParameters {
+			parameters[k] = v
+		}
+	} else {
+		parameters = uriParameters
+	}
+
 	if secretKey == "" {
 		secretKey = inferenceService.ObjectMeta.Annotations[v1beta1.SecretKeyAnnotation]
 	}
+	if secretKey == "" {
+		if t, ok := parameters["type"]; ok {
+			// default key name for the storage type, used for backwards
+			// compatibility with the changes to the storage spec
+			secretKey = fmt.Sprintf("_%s_default", t)
+		}
+	}
```

Comment:
Trying to remember my prior thought process around this... I don't _think_ we want to set it here, rather just leave it empty and let the defaulting logic happen later as part of the general Predictor processing.

One reason for that is that we want to treat the default secret keys slightly differently (as previously discussed) where it's not an error if they don't exist.

We need to produce an error if neither `secretKey` nor type is provided (where type may be implicit from e.g. URI or deprecated `S3` syntax), but I think that can also be common in the later Predictor processing logic.

Do you see any problem with the above approach?

Reply:
(I'm also unsure about the naming of these default entries, but if we are removing that from here it doesn't need to be resolved for this PR!)

Reply:
> just leave it empty and let the defaulting logic happen later as part of the general Predictor processing.
Ah, I thought that the default value for StorageKey would only be injected for an InferenceService, not for a Predictor.

> We need to produce an error if neither secretKey nor type is provided
This would have to come from the Puller, right? To support the custom runtime with no storage helper case, we would not want an error thrown for this in the Controller.

> if we are removing that from here it doesn't need to be resolved for this PR!
So, for this PR, you do not want any default key to be injected, just pass through an empty key? The "later" Predictor processing logic you reference would be in the Puller, right? Is that where you want the default key logic to live?


Agreed about the name of `_%s_default` not being great. That is open for suggestions, maybe something like `_%s_serviceaccount` to tie it back to the SAs, but I don't have a strong suggestion at this point.

Reply:
> This would have to come from the Puller, right? To support the custom runtime with no storage helper case, we would not want an error thrown for this in the Controller.

Good point, yes.

> So, for this PR, you do not want any default key to be injected, just pass through an empty key?

Yes

> The "later" Predictor processing logic you reference would be in the Puller, right? Is that where you want the default key logic to live?

Yes and yes, unless you can think of a major reason for not doing that.